### ISSUE TITLE
AbstractModel `viewUrl`s

### ISSUE THREAD

**Issue by Allcharles** [QutEcoacoustics/workbench-client#772]
Currently, the `viewUrl` method for AbstractModels can only handle setting the url path, and route parameters (by combining the StrongRoute with the routeParams. This is limiting as the model is unable to affect the query parameters of its route. There are two methods of bypassing this problem:

- Create a MenuItemDirective which has the following properties:
  - Accepts in the following input:

  ```typescript
  @Directive({
    selector: "a[menuItem]"
  })
  class MenuItemDirective {
    @Input() menuItemPath: {
      menuItem: NavigatableMenuItem,
      routeParams: RouteParams,
      queryParams: Params
    }
  ```

  - This will function similar to the StrongRouteDirective, with the added benefit of only requiring one input, and will allow interchangable use of `MenuLink`s and `MenuRoute`s
  - This will require adding a default route for external routing similar to the following: https://adrianfaciu.dev/posts/angular-router-external-links/
  - The downside to this solution is that it will require a large chunk of changes to existing code
- Models create copies of the StrongRoute and modify the object to have the correct routeParams and queryParams:
  - Requires minor modification to the StrongRoute class (add`.copy()`, `defaultRouteParams`, `defaultQueryParams`)
  - Requires some changes to angular UI code to change from `<a [routerLink]="model.viewUrl"/>` to `<a strongRoute [strongRoute]="model.viewUrl />"
  - Before

  ```typescript
  public get viewUrl(): string {
    return libraryMenuItem.route.toRouterLink({projectId: this.id});
  }
  ```

  - After

  ```typescript
  public get viewUrl(): StrongRoute {
    const route = libraryMenuItem.route.copy();
    route.defaultRouteParams = (params) => ({customValue: true});
    route.defaultQueryParams = (params) => ({customValue: true});
    return route;
  }
  ```

- Rewrite how StrongRoute works so that we can completely bypass this issue with `RouterLinkWithHref` and `StrongRouteDirective` requiring multiple inputs

**Comment by Allcharles** [comment]
Currently this solution cannot be implemented as it will heavily interfere with the changes proposed during the migration. This will have to be implemented afterwards

**Comment by atruskie** [comment]
I'm unclear on the problem this is trying to solve. Can you show me an example of a failing case at the moment?

**Comment by Allcharles** [comment]
The `Tag` model is the main example which comes to mind. A common route for it is to link to: `/library?reference=all&tagsPartial=:tagText`. That route cannot be extracted to the tag model (using the current implementation), and would instead need to be implemented in any component which wishes to route to that page.

**Comment by Allcharles** [comment]
Another more common example is the `Site` models `playUrl` (which will be moved to the `AudioEvent` model instead). The audio event in question usually specifies a start and end time in seconds `/listen/:audioRecordingId?start=:start&end=:end`. These values are query parameters which cannot be used in the current setup.

**Comment by atruskie** [comment]
So the tag example is... Not great. 

But why can't qsp be used? I forget what we called the method but strong routes have an overload that take qsp and route parameters? All the information in the play link example is available in the audio event model.

(Also site can have a play url, it's just it has no offsets)

(Also models can have multiple url forming functions, which can take parameters, but the viewURL url should always point the the "show" page. We don't have a show page yet for tags, hence the example is not great).

**Comment by Allcharles** [comment]
`StrongRoute` does have qsp, however it only works with the `StrongRouteDirective`. The second option I presented is the closest to what you are suggesting. The reason I can't just set the qsp on the `StrongRoute` is because that will change the qsp for all references to that `StrongRoute`. Hence the need for a `copy` function. I can't just return the string value of the StrongRoute because the angular router will ignore the query parameters.

**Comment by atruskie** [comment]
>  I can't just return the string value of the StrongRoute because the angular router will ignore the query parameters.

Bloody angular. 

> The reason I can't just set the qsp on the StrongRoute is because that will change the qsp for all references to that StrongRoute

Why? It really really shouldn't. We ensure the strong route formatting methods are immutable and idempotent.

---

I think we should create a `UrlRouteDirective` used like so:

```html
<a [url]="audioEvent.viewUrl">My Link</a>
```

The `url` `@Input` should accept the same parameters as `router.navigateByUrl` (either a string or an UrlTree) and the directive should look kind of like `RouterLinkWithHref` - as in it updates the `href` attribute when `url` changes and does not activate when modifier keys are held down.

Making this a little more generalised will allow us to navigate with strings (wow, what cutting edge tech 🙄) or we can have our models return `UrlTree`s if we want to avoid the perf hit associated with the decode step required when parsing a string.

**Comment by atruskie** [comment]
As for external links, I believe just assigning to the `href` attribute directly will skip angular's router.

That can be done in the external link menu item template without doing anything fancy

**Comment by Allcharles** [comment]
Inserting private discussion to help track ideas:

@Allcharles 
> Alright, optimistically implementing the solution you would like will take 2 days. Pessimistically 3. The reason is because, at minimum, im going to have to update around 25 pages (almost all forms use a models viewUrl function so they will all be affected). The tests ive been writing to validate navigation is correct works in 1 of 2 ways:
>
> 1. (old) I read the href value of the link and validate that it is correct (this is a fairly naive solution, and isnt the best because of how its implemented)
> 2. (current) I retrieve the RouterLinkWithHref directive from the anchor, and validate the routerLink and queryParams are correct. This is a good solution, but will have to be re-written if i implement the new UrlRouteDirective you are proposing
> 3. (not used) The ideal solution is to click on the link and validate what route the router attempts to navigate to. Its the hardest solution (although i think i know how to do). If i had implemented this solution, the tests would not need to be changed at all
>
>Now as for handling external routes, the reason i wanted to pass NavigableMenuItems is because then i dont need to check the model if they use an internal or external route. This means i a) dont need to update everywhere in the app when i move from an external link to an internal link, b) dont need to know if the link is external or internal. However, the edge case this is dealing with will dissappear once i migrate baw-client so it may be a non-issue

@atruskie 

> Okay, drastically reduce the scope of this. You only need to make the change for viewUrls that use query Params. Even then, make the directive, test it works, then you can bulk find and replace.
> As for testing: validation of the href is good. Keep doing that. Testing route navigation should be a solvable problem, should be plenty of examples already on the interwebs. BUT don't start doing it unless it's easy.
> External: so the navigable menu items should be an entirely different problem. You shouldn't need to update links when you switch from one to other... They already use different templates. Maybe elucidate the problem more?
>     https://angular.io/guide/testing-components-scenarios#routing-component
>     
> 
> i think the confusion here is: you're envisioning this:
> 
> 
> ```<a [menuItem]="someMenuItem"> A LINK </a>```
> 
> And we already have the menu component, which switches on link type. The real issue seems to passing parameters to the strongroute directive that's already in the link component https://github.com/QutEcoacoustics/workbench-client/blob/d9aff12713ae07491c7f6e42a242b2f1bc82083a/src/app/components/shared/menu/internal-link/internal-link.component.ts#L22
> 
> So that to me says MenuInternalLinkComponent just needs to accept more inputs
> 
> and those inputs should be optional
> as for external link component: it accepts a href already, so any normal string formatter URL (i.e. from one of those model methods) will work
> https://github.com/QutEcoacoustics/workbench-client/blob/d9aff12713ae07491c7f6e42a242b2f1bc82083a/src/app/components/shared/menu/external-link/external-link.component.ts#L19

@Allcharles 

> > https://angular.io/guide/testing-components-scenarios#routing-component
> 
> The actual solution is here: https://angular.io/guide/testing-components-scenarios#components-with-routerlink, which appears to mirror my current solution. This will have to be rewritten if/when i change how the routing works to a custom directive
> 
> 
> > So that to me says MenuInternalLinkComponent just needs to accept more inputs
> 
> 
> This is an extremely confusing statement. MenuInternalLinkComponent is only controllable from the SecondaryMenuComponent and ActionMenuComponent. All of which are in a completely separate router outlet and can only get data through the route data. Not sure what input bypasses the issues we are currently having as the route data must either be generated during compile time, or before a navigation event. Both of which are not exactly viable for the current problem as the link is generated at runtime and would require a page refresh to transfer the data to the MenuInternalLinkComponent
> 
> 
> > i think the confusion here is: you're envisioning this:
> > 
> >
> > ```<a [menuItem]="someMenuItem"> A LINK </a>```
> 
> 
> 
> I was proposing this as a potential solution as it would allow me to use menuLinks and menuRoutes interchangably. This was important when the models had viewUrls which were sometimes internal, and sometimes external links, with no way to check which it was without actually checking the source code. However, once migration of the baw client is complete, this will no longer be an issue because it will all be internal links
> 
> > > The reason I can't just set the qsp on the StrongRoute is because that will change the qsp for all references to that StrongRoute
> >
> > Why? It really really shouldn't. We ensure the strong route formatting methods are immutable and idempotent.
> 
> Individual StrongRoutes are effectivally global variables. I define a homeMenuItem, which is exported and imported everywhere throughout the app. Any changes to that variable will directly affect all other references to it. My proposed solution is to make a copy of the StrongRoute where required such that it will not affect anything else 
> 
> > I think we should create a UrlRouteDirective used like so:
> > 
> > ```<a [url]="audioEvent.viewUrl">My Link</a>```
> > 
> > The url @Input should accept the same parameters as router.navigateByUrl (either a string or an UrlTree) and the directive should look kind of like RouterLinkWithHref - as in it updates the href attribute when url changes and does not activate when modifier keys are held down.
> > Making this a little more generalised will allow us to navigate with strings (wow, what cutting edge tech 🙄) or we can have our models return UrlTrees if we want to avoid the perf hit associated with the decode step required when parsing a string.
> 
> This is a fairly decent idea, and i like it. Using UrlTrees would be an excellent way of killing two birds with one stone
> 
> And for the final thought, almost none of this entire problem affects the MenuComponent in any way. The secondary and action menu work fine, it is only the links inside the page component which use the viewUrl of a model which are the problem. Currently the following is the only way to properly pass query parameters to the models viewUrl:
> ```<a [routerLink]="model.viewUrl" [queryParams]="{siteId: model.id}"/>```
> 
> This is obviously a problem as i would prefer the models viewUrl passes the necessary query params as well. It should instead be (as you proposed):
> ```<a [url]="model.viewUrl"/>```
> 
> One again, this will only affect components in the PageComponent and has no impact on the menus. However, implementing this solution will require modifying a fair chunk of components, and their accompanying unit tests which will take some time and might be a bit messy.

**Comment by atruskie** [comment]
You're not changing all routing to use a custom directive. Only routes that can't use routerLink for some reason. 

Re menulink stuff: fine. It still needs more input, the question is from where. Show me an actual example of where it fails. 

Re menu items: I don't get it. If all you want to do is use menu items instead of view model links, why don't you use them?

 Strong route: they should be immutable constructs. Formatting a URL should not affect its internal state - ever. Don't violate those principles you don't have a problem. Do not create a copy method, it is not needed (or if it is you need a better example).

Re page components: if the menulink ain't an issue ignore it. Proposed [uri] directive will deal with links from viewmodels

**Comment by Allcharles** [comment]
> Re menulink stuff: fine. It still needs more input, the question is from where. Show me an actual example of where it fails.

Im not sure what you mean? Are you talking about the `MenuInternalLinkComponent`?

> Re menu items: I don't get it. If all you want to do is use menu items instead of view model links, why don't you use them?

Basically, its an alternative, but it effectively is just a dressed up StrongRoute. The only benefit to using MenuItems instead of StrongRoutes is that i would be able to distinguish between internal and external links. However, as ive stated, this will have limited use once migration is complete so it may not be worth the effort.

> Strong route: they should be immutable constructs. Formatting a URL should not affect its internal state - ever. Don't violate those principles you don't have a problem. Do not create a copy method, it is not needed (or if it is you need a better example).

The issue is that i am not formatting the URL. I am changing the query parameters which are **not part of the URL** but are instead a separate input to the link. Setting the query parameters currently would require modifying the query parameters object inside the StrongRoute similar to the example in the issue:

```typescript
public get viewUrl(): StrongRoute {
  const route = libraryMenuItem.route.copy();
  route.defaultRouteParams = (params) => ({customValue: true});
  route.defaultQueryParams = (params) => ({customValue: true});
  return route;
}
```

A potential solution to this is to use the `UrlTree`'s you have suggested. If the output of `toRouterLink()` (or something similar) was a `UrlTree` i would be free to set the query parameters without modifying the `StrongRoute` a la:

```typescript
public get viewUrl(): UrlTree {
  const route = libraryMenuItem.route.toRouterLink();
  route.queryParams = {customValue: true};
  return route;
}
```

**Comment by Allcharles** [comment]
Final design notes:

```typescript
// UrlTree example
public get viewUrl(): string | UrlTree {
  return listenMenuItem.route.formatUrlTree({id: this.id}, {start: startOffset});
}

// String example
public get viewUrl(): string | UrlTree {
  return listenMenuItem.route.format({id: this.id}, {start: startOffset});
}

@Directive({
  selector: "a"
})
class UriDirective {
  @Input uri: string | UrlTree;

  // Handle uri...
}
```

Method of converting string to UrlTree:

```typescript
const tree: UrlTree = router.parseUrl(stringGoesHere);
```

Example usage:

```html
<!-- This works with string and UrlTree -->
<a [uri]="model.viewUrl">link</a> 
```

**Comment by Allcharles** [comment]
After implementation of this solution was achieved, it was discovered that creating the `formatUrlTree` method would require re-implementing `StrongRoute` as part of a service which produces `StrongRoute`s. Similar to a factory in AngularJS. This would allow the `StrongRoute` to access the `Router` and `ActivatedRoute` DI classes which are needed to create `UrlTree`s.
